### DISCUSSION TITLE
Feedback: Authenticating with a SPA without a relay

### DISCUSSION THREAD

**Discussion by baywet** [community/community#40077] [Apps, API and Webhooks] [answered]
### Select Topic Area

Product Feedback

### Body

I've been trying to authenticate from a Single Page Application to GitHub so my application can run searches in private repositories.
To me this looks like it's not possible without a service relay for the following reasons:
- The authentication service DOES NOT support implicit flow [as clearly stated in the docs](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps).
- The authentication service DOES NOT support [PKCE](https://espressocoder.com/2019/10/28/secure-your-spa-with-authorization-code-flow-with-pkce/).
- The authentication service REQUIRES a client secret to be sent over when requesting an access token.
- The authentication service DOES NOT support [CORS pre-flights](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) (required for a POST request which is not "simple" when in browser on a different origin).

To unblock this very common scenario of a SPA wanting to authenticate to the service to be able to access the API, the service would need the following changes: (1 and 4 and (2 or 3))
1. support CORS pre-flights
2. support implicit flow (bad)
3. support PKCE (better)
4. remove the requirement for client secret on the access_token endpoint

The reasons why I don't want to rely on a service relay for the authentication or to query the API are the following:
- additional engineering cost (the service needs to be built, deployed, monitored...)
- additional hosting costs
- additional security concern (now we need a secret, it needs to be rotated, it can be compromised,...)

And for the time being, at the very least, the docs should mention that PKCE is not supported, as well as CORS, making the SPA scenario impossible without a relay.

Additional resources:
- https://github.com/orgs/community/discussions/24562
- https://github.com/github/docs/issues/22270

**Comment by ossanna16** [comment]
Hi there @baywet and welcome to our community! Thank you for the feedback!

**Comment by gopi-suvanam** [comment]
To add to this..
Device Authentication Flow is supported.. It does not require a relay server, but it is not seamless to use in an SPA because verification_url_complete is not generated.. i.e. users will have to manually enter the characters... if OAuth supports verification_url_complete (with user_code in the link) then the authentication experience can be made seamless even for an SPA.. So either implement:
1. PKCE OR
2. Device flow with user_code in the verification link OR
3. Return auth-token without client-secret (not recommended ..)

**Reply by Evidlo** [reply]
Have you gotten this working?  I get CORS errors:

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://github.com/login/device/code. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.`

**Reply by gopi-suvanam** [reply]
It's blocked by design. 
You will have to build a backend.. I am using firebase auth to handle this.

Alternatively u can use device flow but the user will have to enter extra characters

**Reply by Evidlo** [reply]
The error I pasted above is using the device flow.

I think CORS prevent that from working, too: https://github.com/octokit/authentication-strategies.js/?tab=readme-ov-file#device-authentication

Pretty frustrating.  I think Gitlab supports PKCE so I may just switch to that instead.

**Comment by hpsin** [comment, marked as answer]
Hey folks - we just added PKCE support - https://github.blog/changelog/2025-07-14-pkce-support-for-oauth-and-github-app-authentication/

I'm still working on getting CORS added as well. We don't have movement towards removing the client secret requirement at this time, but I'll continue to look at options.

**Reply by kalzoo** [reply]
@hpsin that's good to hear. Is there something special or different about GitHub's security model that prevents conventional support for such a common authentication flow? Adding OAuth support for GitHub this week, I am surprised to see that it took a decade - from [2015](https://github.com/isaacs/github/issues/330) until just now - to address this but still not add CORS support. Is the idea that a lack of CORS support forces developers not to embed client secrets in frontend apps, and it's the client secret that is seen as essential?

Also, please consider making this limitation more prominent in the documentation - your blog post links to both GitHub apps and OAuth apps, but only one of those pages even mentions CORS and even then it's tucked in the middle and hard to find. It would be useful to state up front on those pages that GitHub has an unusual constraint on the auth flow which requires the involvement of a server. A developer who's worked with OAuth apps before (like me) likely does not assume that's the case for a PKCE flow.

Thanks.

**Reply by kingdudely** [reply]
> I'm still working on getting CORS added as well. 

It's been a year, are you sure?
